lesson11/script.js

```diff
@@ -0,0 +1,50 @@
+function showLoadingImg(){
+    const loadingImg = document.createElement('img');
+    loadingImg.src = "loading-circle.gif";
+    loadingImg.alt = "ローディング画像";
+    document.getElementById('js-loading').appendChild(loadingImg);
+}
+
+function removeLoadingImg(){
+    document.getElementById('js-loading').remove();
+}
+
+async function getData(){
+    showLoadingImg();
+    try{
+        const url = "https://api.json-generator.com/templates/szdgGQcOLXuk/data?access_token=hu4bc7qh9znx2m8f53mn4mz2hryvdntkavwbw8j0";
+        const response = await fetch(url);
+        const json = await response.json();
+        const data = await json.data;
+        return data;
+    }catch(e){
+        document.getElementById('js-list').textContent = "データが取得できませんでした";
+        throw new Error(e);
```

Comment:
Q
Where is this Error thrown?

Reply:
Because I want to stop the subsequent processing when the data cannot be acquired by fecth.
If I used `console.error(e)` here, I can't stop loop process in showList ().

Reply:
It is not desirable to throw an error and not have a place to catch it.

>  I used console.error(e) here, I can't stop loop process in showList ().

I see!
The reason the process does not stop is elsewhere.
Here it is.
```js
  if (!result) {
    return (ul.textContent = "条件に一致するデータがありません");
  }
```

This is a good way to throw an exception, catch it, and stop the process.
```js
async function getData() {
  showLoadingImg();
  try {
    const url = "https://ほげほげ";
    const response = await fetch(url);
    const json = await response.json();
    const data = await json.data;
    //もしレスポンスが.okであれば 定数dataを返す
    if (response.ok) {
      return data;
    }
    //そうでなければ例外処理実行
    throw new Error(`${response.status}:${response.text}`);
  } catch (e) {
    //投げられたエラーはここでキャッチできます。
    document.getElementById("js-list").textContent = e;
    console.error(e);
  } finally {
    removeLoadingImg();
  }
}

async function showList() {
  const result = await getData();
  const fragment = document.createDocumentFragment();
  const ul = document.getElementById("js-list");

  //resultがtrueでなければなにもしない処理に変更しました。
  if (result) {
    for (const list of result) {
      const li = document.createElement("li");
      const a = document.createElement("a");
      const img = document.createElement("img");
      a.textContent = list.text;
      a.href = `/${list.to}`;
      img.src = list.img;
      img.alt = list.alt;
      fragment.appendChild(li).appendChild(a).appendChild(img);
    }
    ul.appendChild(fragment);
  }
}
```
This way, there is no need to check in the render function to see if the process is correct.
This is just a rough example, so please let me know if there is a rougher way to write better!
[codesandbox](https://codesandbox.io/s/js-lesson11-forked-pe0e6l?file=/script.js)

Reply:
Thank you for the detailed explanation and code!
Thanks to that, I got a good understanding of how to handle exceptions.

Here is the fix.
https://github.com/nakagawamai/js-lesson/pull/13/commits/85264a59b176cd1661793137882c9aa612c2c629